### DISCUSSION TITLE
There is an option to allow ITestOutputHelper diagnostic messages to be recorded even when the test is successful?

### DISCUSSION THREAD

**Discussion by lucas75** [xunit/xunit#3029] [Question] [answered]
ITestOutputHelper only works if the test is successful. 

I was wondering:
* There is an option to allow the ITestOutputHelper messages to be included in the report even when the test is successful?
* There is an option to capture the console and record in the report?
* There other mechanisms to write diagnostic messages to the log report?

Is useful to have the captured output in the log reports for some reasons:
* We may start with an Assert.True(true), execute the test, inspect the output and code better assertions.
* When we have a incident in production is useful to inspect the test reports to understand why the assertions didn't catch the specific situation and improve the tests. Half of the equation is the source code that is versioned. But half of the equation is the data which is not versioned.

**Comment by bradwilson** [comment, marked as answer]
> There is an option to allow the ITestOutputHelper messages to be included in the report even when the test is successful?

This is already the case for our native runners and reports (example using v2):

![image](https://github.com/user-attachments/assets/23b36344-8c97-4a9d-8047-5b7eacebd71a)

Third party runners like `dotnet test` have their own rules for what gets output to the console. You can pass `--logger "console;verbosity=normal"` to `dotnet test` and it will show our console output in addition to theirs, which would allow you to see the results on the console.

They do also include the output for passing tests in the TRX report:

![image](https://github.com/user-attachments/assets/888ad1e0-ff50-4222-bc84-b9b045cf8c77)

> There is an option to capture the console and record in the report?

You can opt into this in v3 (added in `0.3.0-pre.28`):

![image](https://github.com/user-attachments/assets/e31b9abf-8773-41cc-b8da-561664c6be8e)

> There other mechanisms to write diagnostic messages to the log report?

We also added the ability to capture `Trace` and `Debug` output in v3 (also added in `0.3.0-pre.28`):

![image](https://github.com/user-attachments/assets/9fd353c2-bd66-4793-8760-7085c221f39f)

(`Debug.WriteLine` will only be captured if you're doing a Debug build; Release builds filter out all calls to `Debug` at compile time.)

**Reply by lucas75** [reply]
Let me see if I understood:

The fact vscode shows the ITestOutputHelper messages when the test failed, but not when the test is successful is a design choice [or bug] of the vscode dotnet test extension authors. This limitation will not exists if I generate the reports though command lines with the commands shown above.

**Reply by bradwilson** [reply]
That is correct.
